### PR TITLE
DAOS-8992 test: launch.py updates for on-site testing

### PR DESCRIPTION
**Pull request by phender**
Adding additional launch.py command line options for on-site testing:
 --disable_stop_daos    disables stopping daos_server and daos_agent
                         services after running a test file
 --extra_yaml           enables specifying an additional yaml file to
                         supplement (override/append) existing test
			 yaml values
 --timeout_multiplier   multiply any timeout value in the test yaml by
                         the specified value

Also enabled verifying test yaml file modification without the need to
remotely contact the hosts specified with the --test_servers or
--test_clients arguments by combining the --list and --modify arguments.

Signed-off-by: Phillip Henderson <phillip.henderson@intel.com>

### CODE REVIEW

**Comment by phender** [comment]
Additional updates incoming:
- CXI support
- supporting non-hexadecimal bdev_list placeholders
- extending server/client lists via the --test_servers/--test_clients arguments

Should the `stoenwall_timer` be included in the `--timeout_multiplier` handling?

**Comment by daltonbohning** [comment]
> Should the `stoenwall_timer` be included in the `--timeout_multiplier` handling?

Generally, I don't think so. Some considerations:
1. Most tests using ior and mdtest stonewall specify it in `flags`. E.g.: `-D 100`: https://github.com/daos-stack/daos/blob/2d210a0cb924c70328d0143e58ce08286ef99963/src/tests/ftest/ior/crash_ior.yaml#L32 (hard to multiply this format)
2. I've seen several tests using stonewall even though the block size isn't large enough to ever reach it. E.g. 100s stonewall with only 3s of actual IO. Increasing stonewall for these cases is likely benign, but it's a gray area. (This is an issue in itself)
3. For performance tests, stonewall needs to be controlled. E.g.: 30s, 60s, 300s. If we multiple stonewall by `timeout_multiplier`, we might get non-normal stonewalls.

File: src/tests/ftest/launch.py [hunk 1]
```diff
@@ -2184,6 +2223,18 @@ def main():
         action="store_true",
         help="when replacing server/client yaml file placeholders, discard "
              "any placeholders that do not end up with a replacement value")
+    parser.add_argument(
+        "-dsd", "--disable_stop_daos",
+        action="store_true",
+        help="disable stopping DAOS servers and clients between running tests")
```
Comment (daltonbohning) [inline src/tests/ftest/launch.py:2229]: Just a design thought...
```suggestion
    parser.add_argument(
        "-sd", "--stop_daos",
        type=bool,
        default=True,
        help="Stop DAOS servers and clients between running tests")
```

File: src/tests/ftest/launch.py [hunk 2]
```diff
@@ -2260,6 +2309,18 @@ def main():
...
+        action="store",
+        help="collect log sizes and report log sizes that go past provided"
+             "threshold. e.g. '-th 5M'"
+             "Valid threshold units are: B, K, M, G, T")
+    parser.add_argument(
+        "-tm", "--timeout_multiplier",
+        action="store",
+        default=None,
+        type=float,
+        help="a multiplier to apply to each timeout value found in the test yaml")
```
Comment (daltonbohning) [inline src/tests/ftest/launch.py:2323]: Minor - Is `action="store"` necessary when both `default` and `type` are supplied?

**Comment by daltonbohning** [comment]
@phender Could we make it more user friendly to view the yaml generated from `--extra_yaml`? For example, in this usage, how can I view the modified yaml?
```
$ ./launch.py --list --modify --extra_yaml my.yaml performance_ior_easy_dfs_sx
Arguments: Namespace(archive=False, clean=False, disable_stop_daos=False, discard=False, extra_yaml='my.yaml', failfast=False, include_localhost=False, insecure_mode=False, jenkinslog=False, list=True, logs_threshold=None, mode='normal', modify=True, nvme=None, process_cores=False, rename=False, repeat=1, sparse=False, tags=['performance_ior_easy_dfs_sx'], test_clients=None, test_servers=None, timeout_multiplier=None, verbose=0, yaml_directory=None)
Testing with PYTHONPATH=/usr/lib64/python2.7/site-packages/mpich:/home/dbohning/daos/install/lib64/python3.6/site-packages:::/home/dbohning/daos/src/tests/ftest/util/apricot:/home/dbohning/daos/src/tests/ftest/util:/home/dbohning/daos/src/tests/ftest/cart/util:/usr/local/lib64/python3.6/site-packages:/usr/local/lib/python3.6/site-packages:/usr/lib64/python3.6/site-packages:/usr/lib/python3.6/site-packages
Running: fault_status
Total test time: 0s
Running avocado -v
Running with Avocado 69.2
Running avocado list --paginator=off --filter-by-tags=performance_ior_easy_dfs_sx ./
Detected tests:
./performance/ior_easy.py
```

**Comment by daltonbohning** [comment]
@phender Another use case
```
$ ./launch.py --list --modify --test_servers node[1-8] performance_ior_easy_dfs_sx
...
./performance/ior_easy.py
Reading ./performance/ior_easy.yaml
Creating copy: /tmp/tmp7t5aijus/performance-ior_easy.yaml
$ stat /tmp/tmp7t5aijus/performance-ior_easy.yaml
stat: cannot stat ‘/tmp/tmp7t5aijus/performance-ior_easy.yaml’: No such file or directory
```

**Review by mjean308 — DISMISSED** [review]

File: src/tests/ftest/launch.py (hunk 1 above)
Comment (daltonbohning) [reply src/tests/ftest/launch.py:2229]: I was thinking something like `--stop_daos true` vs `--stop_daos false`, but I see your point. No problem by me

**Comment by daltonbohning** [comment]
> If you want to have access to the temporary yaml file generated by launch.py after it completes then you need to specify the -y/--yaml_directory <dir> argument - which writes the modified yaml to the specified directory.

Now I do this, but it doesn't seem --extra_yaml was picked up?
```
$ cat my.yaml
pool:
    mode: 999

$ ./launch.py --list --modify --extra_yaml my.yaml --test_servers node[1-8] performance_ior_easy_dfs_sx -y tmp
...
$ cat tmp/performance-ior_easy.yaml | grep mode
  mode: 146
```

**Comment by phender** [comment]
> Now I do this, but it doesn't seem --extra_yaml was picked up?
> 
> ```
> $ cat my.yaml
> pool:
>     mode: 999
> 
> $ ./launch.py --list --modify --extra_yaml my.yaml --test_servers node[1-8] performance_ior_easy_dfs_sx -y tmp
> ...
> $ cat tmp/performance-ior_easy.yaml | grep mode
>   mode: 146
> ```

Yeah, the extra yaml is applied by the avocado command.  I'm investigating if there is a way to somehow get the combined output.

**Review by daltonbohning — DISMISSED** [review]

File: src/tests/ftest/launch.py [hunk 3]
```diff
@@ -61,22 +62,31 @@ def __del__(self):
             rmtree(self.name)
 
 DEFAULT_DAOS_TEST_LOG_DIR = "/var/tmp/daos_testing"
-YAML_KEYS = {
+YAML_KEYS = OrderedDict({
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/launch.py]: Thanks, fixed.

**Review by mjean308 — APPROVED** [review]

**Comment by phender** [comment]
One test failure in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-8229/14/ is a known issue:
- https://daosio.atlassian.net/browse/DAOS-9860

File: src/tests/ftest/util/server_utils_params.py [hunk 4]
```diff
@@ -333,6 +333,9 @@ class PerEngineYamlParameters(YamlParameters):
                 "CRT_SWIM_RPC_TIMEOUT=10"],
             "ofi+verbs": [
                 "FI_OFI_RXM_USE_SRX=1"],
+            "ofi+cxi": [
+                "FI_OFI_RXM_USE_SRX=1",
```
Comment (mjean308) [inline src/tests/ftest/util/server_utils_params.py:337]: the FI_OFI_RXM_USE_SRX is not needed for cxi;  only needed when the provider is <>;ofi_rxm

**Review by daltonbohning — APPROVED** [review]

**Pull request merged**